File: api/filing/models/deduction_80d_model.py
from .api_base_model import ApiBaseModel


class Deduction80DHealthInsuranceModel(ApiBaseModel):
    health_id: int | None = None
    deduction_id: int | None = None
    filing_id: int
    taken_for: str  # "Self" or "Self & Family" or "Parents"
    includes_senior_citizen: bool = False
    policy_number: str | None = None
    health_insurance_premium: float
    insurer_name: str | None = None

    @property
    def itr_insurer_name(self) -> str:
        return ((self.insurer_name or "NA").strip() or "NA")[:125]

    @property
    def itr_policy_no(self) -> str:
        return ((self.policy_number or "").strip() or "0")[:75]

    @property
    def itr_health_ins_amt(self) -> int:
        return int(self.health_insurance_premium or 0)


class Deduction80DPreventiveCheckupModel(ApiBaseModel):
    checkup_id: int | None = None
    deduction_id: int | None = None
    filing_id: int
    taken_for: str  # "Self" or "Self & Family" or "Parents"
    includes_senior_citizen: bool = False
    checkup_amount: float


class Deduction80DMedicalExpenditureModel(ApiBaseModel):
    expenditure_id: int | None = None
    deduction_id: int | None = None
    filing_id: int
    taken_for: str  # "Self" or "Self & Family" or "Parents"
    includes_senior_citizen: bool = False
    expenditure_amount: float


class Deduction80DModel(ApiBaseModel):
    deduction_id: int | None = None
    filing_id: int
    health_insurance: list[Deduction80DHealthInsuranceModel] | None = []
    preventive_checkup: list[Deduction80DPreventiveCheckupModel] | None = []
    medical_expenditure: list[Deduction80DMedicalExpenditureModel] | None = []
# Deduction80DModel

File: web/filing/models/deductions/medical/deduction-80d-model.ts
/**
 * Section 80D Deduction Interfaces (Health Insurance)
 * Mirrors: section_80d in FilingModel
 * 
 * API Response Structure:
 * {
 *   deduction_id, filing_id,
 *   health_insurance: [{ health_id, deduction_id, filing_id, taken_for, includes_senior_citizen, policy_number, health_insurance_premium, insurer_name }],
 *   preventive_checkup: [{ checkup_id, deduction_id, filing_id, taken_for, includes_senior_citizen, checkup_amount }],
 *   medical_expenditure: [{ expenditure_id, deduction_id, filing_id, taken_for, includes_senior_citizen, expenditure_amount }]
 * }
 */

/**
 * Health Insurance Model (Deduction80DHealthInsuranceModel)
 */
export interface Deduction80DHealthInsuranceModel {
  healthId?: number | null
  deductionId?: number | null
  filingId: number
  takenFor: string  // "Self" / "Self & Family" / "Parents"
  includesSeniorCitizen: boolean
  policyNumber?: string | null
  healthInsurancePremium: number
  insurerName?: string | null
}

/**
 * Preventive Checkup Model (Deduction80DPreventiveCheckupModel)
 */
export interface Deduction80DPreventiveCheckupModel {
  checkupId?: number | null
  deductionId?: number | null
  filingId: number
  takenFor: string  // "Self" / "Self & Family" / "Parents"
  includesSeniorCitizen: boolean
  checkupAmount: number
}

/**
 * Medical Expenditure Model (Deduction80DMedicalExpenditureModel)
 */
export interface Deduction80DMedicalExpenditureModel {
  expenditureId?: number | null
  deductionId?: number | null
  filingId: number
  takenFor: string  // "Self" / "Self & Family" / "Parents"
  includesSeniorCitizen: boolean
  expenditureAmount: number
}

/**
 * Complete 80D Model
 */
export interface Deduction80DModel {
  deductionId?: number | null
  filingId?: number | null
  healthInsurance?: Deduction80DHealthInsuranceModel[] | null
  preventiveCheckup?: Deduction80DPreventiveCheckupModel[] | null
  medicalExpenditure?: Deduction80DMedicalExpenditureModel[] | null
}




export const INITIAL_80D_FORM_DATA: Deduction80DModel = {
  deductionId: null,
  filingId: 0,
  healthInsurance: [],
  preventiveCheckup: [],
  medicalExpenditure: [],
}

export const INITIAL_HEALTH_INSURANCE_ITEM: Deduction80DHealthInsuranceModel = {
  healthId: null,
  deductionId: null,
  filingId: 0,
  takenFor: '',
  includesSeniorCitizen: false,
  policyNumber: null,
  healthInsurancePremium: 0,
  insurerName: null,
}

export const INITIAL_PREVENTIVE_CHECKUP_FORM_DATA: Deduction80DPreventiveCheckupModel = {
  checkupId: null,
  deductionId: null,
  filingId: 0,
  takenFor: '',
  includesSeniorCitizen: false,
  checkupAmount: 0,
}

export const INITIAL_MEDICAL_EXPENDITURE_FORM_DATA: Deduction80DMedicalExpenditureModel = {
  expenditureId: null,
  deductionId: null,
  filingId: 0,
  takenFor: '',
  includesSeniorCitizen: false,
  expenditureAmount: 0,
}
